checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -252,10 +255,49 @@ protected void refineGTE(
             propagateToOperands((LessThanLengthOf) leftQualifier, right, in, store);
         }
 
+        refineSubtrahendWithOffset(left, right, 0, in, store);
+
         Receiver rightRec = FlowExpressions.internalReprOf(analysis.getTypeFactory(), right);
         store.insertValue(rightRec, atypeFactory.convertUBQualifierToAnnotation(refinedRight));
     }
 
+    /**
+     * Refines the subtrahend in a subtraction which is greater than or equal to a certain offset.
+     * The type of the subtrahend is refined to the type of the minuend with the offset added.
+     *
+     * <p>This is based on the fact that if {@code (minuend - subtrahend) >= offset}, and {@code
+     * minuend + o < l}, then {@code subtrahend + o + offset < l}.
+     *
+     * <p>If gtNode is not a subtraction, the method does nothing.
+     *
+     * @param gtNode the node that is greater or equal to the offset
+     * @param offsetNode a node part of the offset
+     * @param offsetVal a constant part of the offset
```

Comment:
I would use a boolean for this, since it only seems to take on the value of 0 or 1 depending on whether the refinement is being applied in a gt/gte node.